### PR TITLE
try simplifying random.multivariate_normal api

### LINKED ISSUE #1384: Multivariate Normal
Hi all,

I noticed that sampling from predefined distributions works a bit different in jax than it does in numpy. 

Is it possible to use the jax.random module to sample from a multivariate normal distribution (with dense covariance matrix)? If not, is there any other way?

So far I only managed to sample from one with a unit variance, using jax.random.normal.

Cheers,

### PR DESCRIPTION
**Pull request by mattjj**
Standardize the behavior of `shape` arguments in `jax.random`. Also simplify the rank-polymorphic behavior of `jax.random.multivariate_normal`.

The basic idea is that, [following NumPy](https://docs.scipy.org/doc/numpy/reference/random/generated/numpy.random.Generator.normal.html) and my own guesses as to what's least surprising, the shape of a randomly-sampled array is determined by either the `shape` argument to the sampler function or the shape of any parameter arrays. (If the `shape` argument is provided and the parameters have non-scalar shapes, then they must be consistent.)

Some sampler functions don't take any parameters: `uniform`, `normal`, `cauchy`, `exponential`, `gumbel`, `laplace`, `logistic`. (We're following the convention of not including scale parameters.) For these, the `shape` argument determines the result shape and has a default value of `()`.

Other sampler functions _corresponding to random variables with a scalar range_ take one or two parameter arguments (`bernoulli`, `beta`, `gamma`, `pareto`, `t`). We could require these arguments to be scalars, and require users to `vmap` (over key arrays, parameter arrays, or combinations of the two) to produce vectors or arrays of samples. **But I chose not to require using `vmap` because that makes the common use case less practical.** As a result of that choice, we need to fix a convention for how the shapes of array parameters affect the result shape, especially if a `shape` argument is specified. [Following NumPy](https://docs.scipy.org/doc/numpy/reference/random/generated/numpy.random.Generator.normal.html), I implemented this convention: the default value of `shape` is `None` and that indicates that the result shape is computed by broadcasting together the parameter shapes. If a non-`None` value of `shape` is provided, then it determines the result shape (and parameter shapes must be broadcast-compatible with it).

One more detail needs to be worked out, which is how to handle the case of random variables with vector ranges (`dirichlet` and `multivariate_normal`). In particular: does `shape` specify the batch shape or the full result shape? Concretely, to sample 100 times from a 3D multivariate normal, do we write `random.multivariate_normal(key, mu, cov, shape=(100,))` or `random.multivariate_normal(key, mu, cov, shape=(100, 3))`? [Following NumPy](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.random.multivariate_normal.html) (and TensorFlow, and Pyro, and convenience) I followed the former: the `shape` argument specifies the _batch shape_.

_Original comment:_

@jekbradbury @aeftimia @fehiepsi  Building directly on #1419, this is an experiment in simplifying `random.multivariate_normal`, basically to remove all the rank-polymorphic behavior (which can be implemented using `vmap` instead).

Thoughts?

fixes #1384

### CODE REVIEW

**Review by jekbradbury — APPROVED** [review]
Shouldn’t the docstring be on `multivariate_normal` and not the underscored function?

**Review by fehiepsi — COMMENTED** [review]
+1 for removing the ambiguous `scalar`, `diagonal` covariance matrix (users should use `normal` sampler for those cases instead). :D 

I think it should be fine to support batching without having to rely on `vmap`. How about the following code snippet?
```
assert np.ndim(mean) >= 1
assert np.ndim(cov) >= 2
n = mean.shape[-1]
assert cov.shape[-2:] == (n, n)
batch_shape = broadcast_shapes(np.shape(mean)[:-1], np.shape(cov)[:-2], shape)
chol_factor = cholesky(cov)
normal_samples = normal(key, batch_shape + (n, 1), dtype)
return np.matmul(chol_factor, normal_samples)[..., 0] + mean
```

**Comment by mattjj** [comment]
Thanks for those catches and suggestions!

@fehiepsi I'm split on whether to follow that NumPy-like rank-polymorphism convention, which is how I'd write things with vanilla NumPy, versus trying to be bold and rely on `vmap` for everything. I just noticed that your `dirichlet` code (the only other non-scalar / non-vectorized sampler) follows the NumPy-like shape-suffix convention.

I think I will change it to the NumPy-like convention, because having some vectorization built into primitives is convenient for expressiveness. If we don't do that, pretty much every user would end up using `vmap` here, and that's just redundant.

**Comment by mattjj** [comment]
@fehiepsi actually, I think the shape behavior in `beta` (and `dirichlet`) are a bit inconsistent and surprising.

```python
In [8]: random.beta(key, np.ones((2, 3)), np.ones((2, 3)), shape=(2, 3)).shape
Out[8]: (2, 3)

In [9]: random.beta(key, np.ones((2, 3)), np.ones((2, 3)), shape=(1, 3)).shape
Out[9]: (1, 2, 3)
```

We need to figure out a clear and consistent way for the output's shape to be computed from the input shapes of the parameters and the `shape` argument.

Here are some options:

1. The `shape` argument determines the output shape. In the above example, the second line would result in an error. More generally, the input arrays would be broadcast against each other, and the result should be a suffix of the value of the `shape` argument.
2. The `shape` argument specifies a prefix on the output shape, basically saying "add these leading dimensions". In the above example, the result shape would become `(1, 3, 2, 3)`.

Are there others? Which do you like better? (I have a slight lean towards the former right now...)

**Comment by mattjj** [comment]
I think the second option above would fit with a comment in the `dirichlet` docstring ("representing the batch shape"). But I'm liking more and more the idea that the shape specifies the full result shape, not just part of the result shape.

**Comment by fehiepsi** [comment]
Hi Matt, I think the issue at `beta` sampler is:
```
np.broadcast_to(np.ones((2, 3)), (1, 3)).shape
```
will return `(1, 2, 3)`, which is an unexpected behavior of `np.broadcast_to`. The expected behavior of `random.beta(key, np.ones((2, 3)), np.ones((2, 3)), shape=(1, 3)).shape` is to throw an error because parameters can not be broadcasted to the shape `(1, 3)`. To me, it is good to have `shape` representing either full shape (=batch_shape + last_dim for multivariate distributions) or batch shape, as long as it is consistent. The only advantage of having `shape` representing batch shape is to match NumPy's behavior (for `size` arg). If I want to take 1000 samples from 3-D MVN distribution, either specify `shape=(1000, 3)` or `shape=(1000,)` is fine to me. The former makes the shape of the output more explicitly, while the latter is a bit less redundant.

My code snippet above will make the behaviour inconsistent, I meant
```
batch_shape = shape or broadcast_shapes(np.shape(mean)[:-1], np.shape(cov)[:-2])
```
instead of
```
batch_shape = broadcast_shapes(np.shape(mean)[:-1], np.shape(cov)[:-2], shape)
```

If you want to change `shape` meaning to full result shape, then the code should be
```
batch_shape = shape[:-1] if shape else broadcast_shapes(np.shape(mean)[:-1], np.shape(cov)[:-2])
```

**Comment by mattjj** [comment]
Awesome, good thinking. And good example (drawing 1000 samples from an MVN) to distinguish the two good options, namely shape being batch shape and shape being full shape.

Sounds like treating `shape` as batch shape is the most common, as I'm told it's what TFP, Pyro, and [even NumPy](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.random.multivariate_normal.html) does. I think we should prefer that behavior on the grounds that it'll be the least surprising.

**Comment by mattjj** [comment]
I updated the comment at the top of this thread to reflect the place I arrived at. @jekbradbury @fehiepsi if you have time, can you read over that and lmk what you think?

**Comment by mattjj** [comment]
(I should add tests for this shape logic, but I need to move on to other things for now!)

**Comment by fehiepsi** [comment]
I really like the change: `shape=()` -> `shape=None`. Approve! :D

**Review by jekbradbury — APPROVED** [review]
I endorse both the result and the reasoning :)

**Comment by mattjj** [comment]
Thanks, all!
